domain/bookings.py

```diff
@@ -25,3 +48,56 @@ def filter_bookings_to_compute_remaining_stock(stock: Stock) -> Iterator:
                                 and b.dateUsed
                                 and b.dateUsed >= stock.dateModified),
                   stock.bookings)
+
+
+def generate_qr_code(booking: Booking) -> str:
+    qr = qrcode.QRCode(
+        version=QR_CODE_VERSION,
+        error_correction=qrcode.constants.ERROR_CORRECT_Q,
+        box_size=QR_CODE_BOX_SIZE,
+        border=QR_CODE_BOX_BORDER,
+    )
+
+    offer_formula = ''
+    if booking.stock.offer.type == str(EventType.CINEMA):
+        offer_formula = 'PLACE'
+    if booking.stock.offer.type == str(ThingType.CINEMA_ABO):
+        offer_formula = 'ABO'
+
+    offer_type = 'EVT' if booking.stock.offer.isEvent else 'BIEN'
+    offer_date_time = booking.stock.beginningDatetime if booking.stock.beginningDatetime else ''
+    product_extra_data = booking.stock.offer.product.extraData
+    product_isbn = ''
+    if product_extra_data:
+        if 'isbn' in product_extra_data:
+            product_isbn = product_extra_data['isbn']
+
+    data = f'PASSCULTURE:{QR_CODE_PASS_CULTURE_VERSION};' \
+           f'TOKEN:{booking.token};' \
+           f'EMAIL:{booking.user.email};' \
+           f'OFFERID:{humanize(booking.stock.offer.id)};' \
+           f'OFFERNAME:{booking.stock.offer.name};' \
+           f'VENUE:{booking.stock.offer.venue.name};' \
+           f'TYPE:{offer_type};' \
+           f'PRICE:{booking.stock.price};' \
+           f'QTY:{booking.quantity};'
+
+    if offer_formula != '':
+        data += f'FORMULA:{offer_formula};'
+
+    if offer_date_time != '':
+        data += f'DATETIME:{offer_date_time};'
+
+    if product_isbn != '':
+        data += f'EAN13:{product_isbn};'
+
+    qr.add_data(data)
+    image = qr.make_image(fill_color='black', back_color='white')
+    return _convert_to_base64(image)
+
+
+def _convert_to_base64(image: Image) -> str:
+    image_as_bytes = io.BytesIO()
+    image.save(image_as_bytes)
+    image_as_base64 = base64.b64encode(image_as_bytes.getvalue())
+    return f'data:image/png;base64,{str(image_as_base64, encoding="utf-8")}'
```

Comment:
```suggestion
    return f'{str(image_as_base64, encoding="utf-8")}'
```
`data:image/png;base64,` étant un détail d'implémentation, il faudrait plutôt le mettre côté front.

Reply:
👍